### PR TITLE
Make captcha and password optional for external accounts

### PR DESCRIPTION
**Pull request by coolaj86**
Replaces https://github.com/go-gitea/gitea/pull/5029/files

* [x] Make CAPTCHAs optional for External Accounts (i.e. Github, OpenID Connect, etc)
  * [x] Add `REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA`
* [x] Make Passwords optional for External Accounts (i.e. Github, OpenID Connect, etc)
  * [x] Add `REQUIRE_EXTERNAL_REGISTRATION_PASSWORD`
* [x] **SECURITY**:
  * [x] This increases security by allowing users to eliminating the weakest link (password) on an otherwise secure account.
  * [x] This makes it more convenient for users to make the secure choice (using a secure external account) than the insecure choice (waiting for an email from haveibeenpwned.com about their shared password being exploited)

This is part of the group of work for collective UX improvement and account security related to passwords and secondary accounts.

### CODE REVIEW

File: modules/setting/service.go [hunk 1]
```diff
@@ -60,6 +62,8 @@ func newService() {
 	Service.EnableReverseProxyAutoRegister = sec.Key("ENABLE_REVERSE_PROXY_AUTO_REGISTRATION").MustBool()
 	Service.EnableReverseProxyEmail = sec.Key("ENABLE_REVERSE_PROXY_EMAIL").MustBool()
 	Service.EnableCaptcha = sec.Key("ENABLE_CAPTCHA").MustBool(false)
+	Service.RequireExternalRegistrationCaptcha = sec.Key("REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA").MustBool()
```
Comment (lafriks) [inline modules/setting/service.go]: I think it should default to `Service.EnableCaptcha` to not change current behavior:
```suggestion
	Service.RequireExternalRegistrationCaptcha = sec.Key("REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA").MustBool(Service.EnableCaptcha)
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -211,6 +214,8 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `ENABLE_REVERSE_PROXY_EMAIL`: **false**: Enable this to allow to auto-registration with a
    provided email rather than a generated email.
 - `ENABLE_CAPTCHA`: **false**: Enable this to use captcha validation for registration.
+- `REQUIRE_EXTERNAL_REGISTRATION_CAPTCHA`: **false**: Enable this to force captcha validation
```
Comment (lafriks) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:231]: it should probably be noted that default depends on `ENABLE_CAPTCHA`

**Comment by solderjs** [comment]
Updated and rebased.

File: routers/user/auth.go [hunk 3]
```diff
@@ -776,10 +782,13 @@ func LinkAccountPostSignIn(ctx *context.Context, signInForm auth.SignInForm) {
...
 // LinkAccountPostRegister handle the creation of a new account for an external account using signUp
 func LinkAccountPostRegister(ctx *context.Context, cpt *captcha.Captcha, form auth.RegisterForm) {
+	// TODO Make insecure passwords optional for local accounts also,
+	//      once email-based Second-Factor Auth is available
+	ctx.Data["AllowPassword"] = setting.Service.RequireExternalRegistrationPassword && !setting.Service.AllowOnlyExternalRegistration
 	ctx.Data["Title"] = ctx.Tr("link_account")
 	ctx.Data["LinkAccountMode"] = true
 	ctx.Data["LinkAccountModeRegister"] = true
-	ctx.Data["EnableCaptcha"] = setting.Service.EnableCaptcha
+	ctx.Data["EnableCaptcha"] = setting.Service.RequireExternalRegistrationCaptcha
```
Comment (kolaente) [inline routers/user/auth.go]: as above

File: routers/user/auth.go [hunk 4]
```diff
@@ -116,6 +118,7 @@ func SignIn(ctx *context.Context) {
 		return
 	}
 
+	ctx.Data["AllowPassword"] = true
```
[the remark this answers is not in the document]
Comment (coolaj86) [reply routers/user/auth.go]: It's the sign-up with password page. It doesn't make sense to get to that page unless it's true.

File: routers/user/auth.go [hunk 5]
```diff
@@ -658,9 +662,10 @@ func oAuth2UserLoginCallback(loginSource *models.LoginSource, request *http.Requ
 
 // LinkAccount shows the page where the user can decide to login or create a new account
 func LinkAccount(ctx *context.Context) {
+	ctx.Data["AllowPassword"] = setting.Service.RequireExternalRegistrationPassword && !setting.Service.AllowOnlyExternalRegistration
 	ctx.Data["Title"] = ctx.Tr("link_account")
 	ctx.Data["LinkAccountMode"] = true
-	ctx.Data["EnableCaptcha"] = setting.Service.EnableCaptcha
+	ctx.Data["EnableCaptcha"] = setting.Service.RequireExternalRegistrationCaptcha
```
[the remark this answers is not in the document]
Comment (coolaj86) [reply routers/user/auth.go]: Okay.

**Comment by lafriks** [comment]
@solderjs are you planing to continue work on this?

**Comment by zeripath** [comment]
I think `ALLOW_PASSWORD` is probably not the correct name for this template data switch - perhaps `SHOW_PASSWORD_FIELD`?

**Comment by zeripath** [comment]
OK I've taken the liberty of updating and addressing @kolaente's concerns

**Review by techknowlogick — APPROVED** [review]

File: routers/user/auth.go [hunk 6]
```diff
@@ -854,30 +861,49 @@ func LinkAccountPostRegister(ctx *context.Context, cpt *captcha.Captcha, form au
...
-	if len(strings.TrimSpace(form.Password)) > 0 && len(form.Password) < setting.MinPasswordLength {
-		ctx.Data["Err_Password"] = true
-		ctx.RenderWithErr(ctx.Tr("auth.password_too_short", setting.MinPasswordLength), tplLinkAccount, &form)
-		return
+	if setting.Service.AllowOnlyExternalRegistration || !setting.Service.RequireExternalRegistrationPassword {
+		// Generating a random password a stop-gap shim to get around the password requirement.
+		// Eventually the database should be changed to indicate "Second Factor"-enabled accounts
+		// (accounts that do not introduce the security vulnerabilities of a password).
+		// If a user decides to circumvent second-factor security, and purposefully create a password,
+		// they can still do so using the "Recover Account" option.
```
Comment (zeripath) [inline routers/user/auth.go:886]: This is not ideal. We currently don't allow users to login with an empty password. So we should just set the password to be empty.

**Comment by zeripath** [comment]
I've also removed the random password as an empty password is just as good.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@lafriks need your approval.

File: modules/auth/user_form.go [hunk 7]
```diff
@@ -79,7 +79,7 @@ func (f *InstallForm) Validate(ctx *macaron.Context, errs binding.Errors) bindin
 type RegisterForm struct {
 	UserName           string `binding:"Required;AlphaDashDot;MaxSize(40)"`
 	Email              string `binding:"Required;Email;MaxSize(254)"`
-	Password           string `binding:"Required;MaxSize(255)"`
+	Password           string `binding:"MaxSize(255)"`
```
Comment (zeripath) [inline modules/auth/user_form.go:82]: Does this change need a migration?

Comment (techknowlogick) [reply modules/auth/user_form.go:82]: I'm pretty sure no, because this is a form validation not a struct that xorm uses.

Comment (zeripath) [reply modules/auth/user_form.go:82]: ah sorry of course!

**Comment by solderjs** [comment]
Is there anything else that this needs presently? It's a holiday today so I have some free cycles.

**Comment by solderjs** [comment]
> I've also removed the random password as an empty password is just as good.

This is not particularly true.

It is _very_ likely that a programmer will make a mistake (whether already in the past, or in the future), in which a comparison is made between some API or Form input and the password, not checking to see if the password is blank.

It is very, very _unlikely_ that the programmer will accidentally check the password against a random string.

**Comment by solderjs** [comment]
> I think `AllowPassword` is probably not the correct name for this template data switch - perhaps `ShowPasswordFormField`?

`AllowPassword` was chosen intentionally because passwords represent one of the greatest security vulnerabilities.

My commits have been aimed at providing authentication patterns that are both **more secure** and **easier to use** than the default password authentication. My hope is that one day passwords will be disabled by default and you will have to specifically opt-in to using a password with some sort of "warning: you're opening up your gitea instance and all of your users to pwnage". Hence `AllowPassword` as in "Allow this bad thing, which normally shouldn't be allowed".

That said, call it what you will.

**Comment by zeripath** [comment]
Hi @solderjs I understand what you're saying about comparing empty passwords to empty strings - but we've already had to deal with this problem and have already fixed it. So whilst your concern would be correct - we've got the fix in.

Now because of our logic - if the password is empty we can assume that the password is not set. Thus we can migrate to remove the local password in future easily - because an empty password is an unset one. Yes, I agree this is not ideal and I would prefer to do this properly - however we cannot do that at this point in 1.9. If you stick a random password in - I won't be able to do that migration in future. 

Removing the special status of the local db password is on my personal list of things to do, but my list is long. The technical debt of Gogs is very large IMHO. 

---

Now, regarding `AllowPassword` marker in the templates. It just didn't make sense to me when reading the templates. I completely understand what you're aiming for, but we have to make sure that reading the templates can be done without that global understanding - TBH `DisablePassword` isn't perfect in that regard either. 

--- 

I do appreciate what you're aiming for though.

**Comment by solderjs** [comment]
👍

**Review by lafriks — APPROVED** [review]

**Pull request merged**